File: algorithms/java/src/_163_MissingRanges/_163_MissingRanges.java
package _163_MissingRanges;

import java.util.ArrayList;
import java.util.List;

public class _163_MissingRanges {
	public static void main(String[] args) {
	    _163_MissingRanges_ so = new _163_MissingRanges_();

	}
}

class _163_MissingRanges_ {

}

 class Solution1 {
	/*
	 * @param nums: a sorted integer array
	 * @param lower: An integer
	 * @param upper: An integer
	 * @return: a list of its missing ranges
	 */

	private void addRange(List<String> res, long st, long ed) {
		if (st > ed)
			return;
		if (st == ed) {
			res.add(st + "");
			return;
		}
		res.add(st + "->" + ed);
	}

	public List<String> findMissingRanges(int[] nums, int lower, int upper) {
		// write your code here
		List<String> res = new ArrayList<String>();
		if (nums == null || nums.length == 0) {
			addRange(res, (long)lower, (long)upper);
			return res;
		}

		addRange(res, (long)lower, (long)nums[0] - 1);

		for (int i = 1;i < nums.length;i++) {
			addRange(res, (long)nums[i - 1] + 1, (long)nums[i] - 1);
		}

		addRange(res, (long)nums[nums.length - 1] + 1, (long)upper);

		return res;
	}
}

File: algorithms/cpp/_163_MissingRanges/_163_MissingRanges.cpp

# include <iostream>

using namespace std;

int main() {
    cout << "new file!" << endl;
    return 0;
}

class Solution1 {
public:
    /**
     * @param nums a sorted integer array
     * @param lower an integer
     * @param upper an integer
     * @return a list of its missing ranges
     */
    string getString(long long st,long long ed) {
        if (st > ed) {
            return "";
        }
        if (st == ed) {
            return to_string(st);
        }
        return to_string(st) + "->" + to_string(ed);
    }

    vector<string> findMissingRanges(vector<int>& nums, int lower, int upper) {
        // Write your code here
        vector<string> ans;
        string tmp;

        vector<long long> nums2;
        for (int i = 0; i < nums.size(); i++) {
            nums2.push_back(nums[i]);
        }

        if (nums2.empty()) {
            tmp = getString(lower, upper);
            if (tmp != "") {
                ans.push_back(tmp);
            }
            return ans;
        }
        //////////////////////////////////////////////////////////

        tmp = getString(lower, nums2[0] - 1);
        if (tmp != "") {
            ans.push_back(tmp);
        }

        for (int i = 1; i < nums2.size(); i++) {
            tmp = getString(nums2[i-1] + 1, nums2[i] - 1);
            if (tmp != "") {
                ans.push_back(tmp);
            }
        }

        tmp = getString(nums2.back() + 1, upper);
        if (tmp != "") {
            ans.push_back(tmp);
        }

        return ans;
    }
};
